src/Cli/dotnet/commands/dotnet-new/MSBuildEvaluation/MSBuildEvaluator.cs

```diff
@@ -0,0 +1,201 @@
+                    //was restored or not, so it is not checked.
+                    _logger?.LogDebug("Project is non-SDK style, cannot evaluate restore status, succeeding.");
+                    return NonSDKStyleEvaluationResult.CreateSuccess(projectPath, evaluatedProject);
+                }
+
+                //For SDK-style project, if the project was restored "RestoreSuccess" property will be set to true.
+                if (!evaluatedProject.GetProperty("RestoreSuccess")?.EvaluatedValue.Equals("true", StringComparison.OrdinalIgnoreCase) ?? true)
+                {
+                    _logger?.LogDebug("Project is not restored, exiting.");
+                    return MSBuildEvaluationResult.CreateNoRestore(projectPath);
+                }
+
+                //If target framework is set, no further evaluation is needed.
+                if (!string.IsNullOrWhiteSpace(targetFramework))
+                {
+                    _logger?.LogDebug("Project is SDK style, single TFM:{0}, evaluation succeeded.", targetFramework);
+                    return SDKStyleEvaluationResult.CreateSuccess(projectPath, targetFramework, evaluatedProject);
+                }
+
+                //If target framework is not set, then presumably it is multi-target project.
+                //If there are no target frameworks, it is not expected.
+                if (targetFrameworks == null)
+                {
+                    _logger?.LogDebug("Project is SDK style, but does not specify the framework.");
+                    return MSBuildEvaluationResult.CreateFailure(projectPath, string.Format(LocalizableStrings.MSBuildEvaluator_Error_NoTargetFramework, projectPath));
+                }
+
+                //For multi-target project, we need to do additional evaluation for each target framework.
+                Dictionary<string, Project?> evaluatedTfmBasedProjects = new Dictionary<string, Project?>();
+                foreach (string tfm in targetFrameworks)
+                {
+                    _logger?.LogDebug("Evaluating project for target framework: {0}", tfm);
+                    evaluatedTfmBasedProjects[tfm] = RunEvaluate(projectPath, tfm);
+                }
+                _logger?.LogDebug("Project is SDK style, multi-target, evaluation succeeded.");
+                return MultiTargetEvaluationResult.CreateSuccess(projectPath, evaluatedProject, evaluatedTfmBasedProjects);
+
+            }
+            catch (Exception e)
+            {
+                _logger?.LogDebug("Unexpected error: {0}", e.Message);
+                return MSBuildEvaluationResult.CreateFailure(projectPath, e.Message);
+            }
+        }
+
+        private static Project RunEvaluate(string projectToLoad, string? tfm = null)
+        {
+            if (File.Exists(projectToLoad))
+            {
+                ProjectCollection projectCollection = new ProjectCollection();
```

Comment:
Is creating a new ProjectCollection required? Generally I'd expect a quick evaluation like this to use the global one, but if isolation is helpful, we should explain why.

Reply:
Changed to use `ProjectCollection.GlobalProjectCollection`

Reply:
Actually tests discovered couple of issues with using `ProjectCollection.GlobalProjectCollection`, so I had to add the checks to see if the project is already loaded.

Couple of questions here:
- what project collection caches? Is it possible to load same project several times with different global properties? or previous project should be unloaded first?
- should we unload the project immediately after evaluation? The comment says so however other uses in sdk never do this.

Reply:
> Is it possible to load same project several times with different global properties? or previous project should be unloaded first?

Yes, you can have multiples of a single project with distinct global properties in a single collection. That will happen automatically in cases like "building a multitargeted project".

> * should we unload the project immediately after evaluation? The comment says so however other uses in sdk never do this.

I suspect much of the SDK can avoid unloading because the process is short-lived, so OS process teardown does all the cleanup required. But this actually raises some interesting questions, since we've moved more execution into the `dotnet` process. @rokonec, do you know if doing evaluations using the GlobalProjectCollection will cause problems with a subsequent build, or with stale cached results from a previous restore "build"?

Reply:
In my test case the evaluation was done from separate processes, i.e. I ran `dotnet new smth` which did evaluation, and then ran `dotnet new smth else` which repeated evaluation. The project didn't change between the run. That's why I added the code that checks if the project is already loaded with same global properties.
Using `GlobalCollection` caches result for different process too, so termination of the process that does evaluation will not remove evaluated project from `GlobalCollection`.

Reply:
We discussed it with @rokonec today
- evaluations are not cached in-between processes
- the cause of above was a bug in the code causing the evaluation happens twice. 
- using new project collection vs global doesn't make much difference perf wise

However, we also discussed that global collection is not thread safe, therefore I'm tempting not to use global collection due to the risk that some other logic in SDK may run evaluation in parallel (though not likely). I switched to using shared non-global project collection in my implementation. This still brings some benefit in case multi-target project, and ensures no other functionality can write to same collection. `MSBuildEvaluator` already ensures that no evaluations are run in parallel.